core/src/main/scala/kafka/log/Log.scala

```diff
@@ -384,7 +429,15 @@ class Log(@volatile var dir: File,
     // checkpoint the recovery point
     stateManager.takeSnapshot()
 
-    val bytesTruncated = segment.recover(stateManager, leaderEpochCache)
+    val bytesTruncated =
+      try segment.recover(stateManager, leaderEpochCache)
+      catch {
+        case e: IndexOffsetOverflowException => {
+          info(s"Splitting segment on offset overflow ${segment.log.file.getAbsolutePath}")
+          Log.splitSegmentOnOffsetOverflow(this, segment)
+          throw e
```

Comment:
I think it's best for callers to retry because we have now have two segments in place of the one we were trying to recover. Callers might maintain some state about segments being that might need to be updated if the segment we are trying to recover gets split into two.

An example is `loadSegmentFiles()` which loops through all .log files, calls `recover()` on a particular segment if the index is corrupt, and finally "publishes" the segment as a valid segment in use. If `recover()` splits the segment into two, `loadSegmentFiles()` somehow needs to know that the segment it passed in no longer exists and is instead replaced by two new ones. To avoid complication from dealing with this, I think a simpler approach is to just fail the entire operation and retry from scratch. This way, `loadSegmentFiles()` will now see the two new segments that were created and will execute without any logic to deal with split segments.